rust/sbd-client/src/send_buf.rs

```diff
@@ -0,0 +1,159 @@
+
+    /// Close the connection.
+    pub async fn close(&mut self) {
+        self.ws.close().await;
+    }
+
+    /// We received a new rate limit from the server, update our records.
+    pub fn new_rate_limit(&mut self, limit: u64) {
+        if limit < self.limit_rate {
+            // rate limit updates are sent on a best effort,
+            // and there are network timing conditions to worry about.
+            // assume we accidentally sent a message while the new limit
+            // was in effect, and accout for that in a brute-force manner.
+
+            let now = self.origin.elapsed().as_nanos() as u64;
+
+            self.next_send_at = std::cmp::max(now, self.next_send_at)
+                + (MAX_MSG_SIZE as u64 * self.limit_rate);
+        }
+        self.limit_rate = limit;
+    }
+
+    /// If we need to wait before taking the next step, this
+    /// returns how long.
+    pub fn next_step_dur(&self) -> Option<std::time::Duration> {
+        let now = self.origin.elapsed().as_nanos() as u64;
+
+        if now - self.last_send >= self.idle_keepalive_nanos {
+            // we need a keepalive now, don't wait
+            return None;
+        }
+
+        if now < self.next_send_at {
+            let need_keepalive_at =
+                self.idle_keepalive_nanos - (now - self.last_send);
+            let nanos =
+                std::cmp::min(need_keepalive_at, self.next_send_at - now);
+            Some(std::time::Duration::from_nanos(nanos))
+        } else {
+            None
+        }
+    }
+
+    /// Call `next_step_dur()` first. If it returns None, or you
+    /// await the Duration returned, call this function to send
+    /// out the next queued block on the low-level websocket.
+    /// Returns true if it did something, false if it did not.
+    pub async fn write_next_queued(&mut self) -> Result<bool> {
+        let now = self.origin.elapsed().as_nanos() as u64;
+
+        // first check if we need to keepalive
+        if now - self.last_send >= self.idle_keepalive_nanos {
+            let mut data = Vec::with_capacity(32);
+            data.extend_from_slice(CMD_FLAG);
+            data.extend_from_slice(b"keep");
+            self.raw_send(now, data).await?;
+            return Ok(true);
+        }
+
+        if self.next_step_dur().is_some() {
+            return Ok(false);
+        }
+
+        if let Some(buf) = self.buf.pop_front() {
+            self.raw_send(now, buf).await?;
+
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
+    /// If our buffer is over our buffer size, do the work to get it under.
+    /// Then queue up data to be sent out.
+    /// Note, you'll need to explicitly call `write_next_queued()` or
+    /// make additional sends in order to get this queued data actually sent.
+    pub async fn send(&mut self, pk: &PubKey, data: &[u8]) -> Result<()> {
+        if data.len() > MAX_MSG_SIZE - 32 {
+            return Err(Error::other("message too large"));
```

Comment:
this could be a distinct error type e.g. `Error::MsgOversized` maybe even with a number indicating the size limit

Reply:
We pre-optimized so many error types in holochain that no-one ever matches on. This library is largely going to be used by a wrapper cryptography library that will always use correct message sizes, so I'm not imagining this is worth spending time on.

Reply:
i created #4 as i still think it could be useful on a `pub fn`